packages/dds/tree/src/feature-libraries/sequence-field/utils.ts

```diff
@@ -774,3 +1062,300 @@ export function getUniqueMoveId<T>(
     }
     return mark.id;
 }
+
+interface DetachedNode {
+    rev: RevisionTag;
+    index: number;
+}
+
+/**
+ * Keeps track of the different ways detached nodes may be referred to.
+ * Allows updating changesets so they refer to a detached node by the details
+ * of the last detach that affected them.
+ *
+ * WARNING: this code consumes O(N) space and time for marks that affect N nodes.
+ * This is code is currently meant for usage in tests.
+ * It should be tested and made more efficient before production use.
+ */
+export class DetachedNodeTracker {
+    private nodes: Map<number, DetachedNode> = new Map();
```

Comment:
What do the keys in this map represent?

Reply:
A node index in the tip state. I added a comment.